### PR TITLE
Week3 step6

### PR DESCRIPTION
**Pull request by hieudan225**
Add deleting all comments and set max comments displaying features
Clean up the code (replace the hard-coded images list with a function iterating through images directory, sanitize the inputs from both server and client side)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.google.sps.data;
+
+public class Comment {
+    
+    private long id;
+    private String name;
+    private String content;
+    private String timestamp;
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: If you store timestamps as strings, determining the format and parsing each time you want to sort things or query for new comments can be a bit annoying. Consider using a time type (I think Java has a LocalDateTime that would be appropriate) or a unix timestamp. Then you would put it into a displayable string at display time.

File: portfolio/src/main/java/com/google/sps/servlets/BackgroundChange.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        if (file.isFile()) {
+            String fileName = file.getName();
+            if (fileName.endsWith(".jpg") || fileName.endsWith(".jpeg")) {
+                String imageURL = "images/" + fileName;
+                this.images.add(imageURL);
+            }
+        }
+    }
+    
+    int randomIndex = (int) (Math.random()*this.images.size());
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/BackgroundChange.java:34]: This is a nice add-on so that it always changes, smart move.

File: portfolio/src/main/java/com/google/sps/servlets/BackgroundChange.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  private int prevIndex = -1;
+
+  /*Assumming background images are .jpg and .jpeg only*/
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    images = new ArrayList<>();
+    File filePath = new File("/home/dhle/step/portfolio/src/main/webapp/images");
+    File[] allFiles = filePath.listFiles();
+    
+    for (File file : allFiles) {
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/BackgroundChange.java:24]: Nice work, this is what I was thinking of in the last revision.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that deletes all comments */
+@WebServlet("/delete-all-comments")
+public class DeleteAllComments extends HttpServlet {
+  public boolean finished = false;
+  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java]: I believe that this is supposed to be a post, generally any method that changes the underlying data is a post.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    long start = System.currentTimeMillis();
+
+    while (System.currentTimeMillis() - start < 1000) {
+        Query query = new Query("comment");
+        query.setKeysOnly();
+        ArrayList<Key> keys = new ArrayList<Key>();
+        for (Entity entity: datastore.prepare(query).asIterable(FetchOptions.Builder.withLimit(128))) {
+            keys.add(entity.getKey());
+        }
+        keys.trimToSize();
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java]: This is probably excessive for your use case, ArrayList is fairly good at determining the necessary memory footprint and you're probably not looking at enough data to make the trim worth it.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    long start = System.currentTimeMillis();
+
+    while (System.currentTimeMillis() - start < 1000) {
+        Query query = new Query("comment");
+        query.setKeysOnly();
+        ArrayList<Key> keys = new ArrayList<Key>();
+        for (Entity entity: datastore.prepare(query).asIterable(FetchOptions.Builder.withLimit(128))) {
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java]: Why 128?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            break;
+        }
+        try {
+            datastore.delete(keys);
+            break;
+        } catch (Throwable ignore) {
+            continue;
+        }
+    }
+    if (this.finished) {
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteAllComments.java]: The finished message is a smart move, since it accounts for the case where the loop times out.

File: portfolio/src/main/java/com/google/sps/servlets/ListComments.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Integer maxComments = Integer.parseInt(request.getParameter("maxComments"));
+    Query query = new Query("comment").addSort("timestamp", SortDirection.DESCENDING);
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Comment> comments = new ArrayList<>();
+    int count = 0;
+    for (Entity entity: results.asIterable()) {
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/ListComments.java]: In general for loops with breaks are less readable than while loops, since you have to parse two statements to understand the conditions under which the loop ends. Would recommend using a while loop over maxComments instead.

File: portfolio/src/main/java/com/google/sps/servlets/NewComment.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+  public static boolean isAlpha(String s) {
+      return s!= null && s.matches("^[a-zA-Z ]*$");
+  }
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String name = request.getParameter("name");
+    String content = request.getParameter("content");
+    
+    if (name.length() < 3 && name.length() > 12) {
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/servlets/NewComment.java]: In general I think that your name validation might be too strict.

* If people put their full names, it could easily be longer than 12 characters 
* Many common names contain apostrophes, dashes, umlauts, etc

I think requiring a length greater than 2 is reasonable (though you got your comparison backwards I think).

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+function isLetter(s) {
+  let alphaExp = /^[a-zA-Z() ]+$/;
+  return alphaExp.test(s);
+} 
+
+async function validateForm() {
+  let name = document.getElementById("name").value;
+  let content = document.getElementById("content").value;
+  if (!isLetter(name)) {
+    alert("JS: Name must alphabet letter only.");
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js]: See above for discussion of name validation.

Generally if you're going to validate inputs, you should do it in one place so that you don't accidentally get them out of sync later. Choose whether you'd rather do it in the front end or backend (either makes sense for different reasons).

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    console.log(imageURL);
+    document.getElementById("bg-image").style.background = "linear-gradient(rgba(0,0,0,0.5),rgba(0,0,0,0.5)), url(" + imageURL +") no-repeat fixed center";
+};
+
+function isLetter(s) {
+  let alphaExp = /^[a-zA-Z() ]+$/;
+  return alphaExp.test(s);
+} 
+
+async function validateForm() {
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js]: This name is a bit of a misnomer: It not only validates the form, but also saves the data if validation passes. validateAndSave or something similar might be more appropriate. If you decide to put data checks on the backend, saveComment might be a good choice.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    const params = new URLSearchParams();
+    params.append('name', name);
+    params.append('content', content);
+    await fetch('/new-comment', {method: 'POST', body: params});
+    location.reload();
+  }
+}
+
+/*Handle onload for comment section in intro.html*/
+async function setMaxComments() {
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js:9]: I'm not a JavaScript expert, but I feel like you might have some excessive async functions throughout this file. Basically what many of these functions do are some quick operations like getting parameters and then block on the slowest operation (the get/post). As a result, you don't get any benefit from the asynchronous behavior that I can see, since you're still waiting for each operation in turn. However, you've taken on the complexity of asynchronous behavior into your error reporting.

In summary, think about what value this asynchronous behavior is bringing to you. Is that value worth the resulting complexity? I'll do some more research on what's recommended here as well and get back to you.

File: walkthroughs/week-3-server/examples/subtraction-game/src/main/webapp/index.html [hunk 13]
```diff
@@ -10,7 +10,7 @@ <h1>The Subtraction Game</h1>
     <p>Take turns subtracting 1, 2, or 3 from the total.<p>
     <p>The player who reaches 0 wins!</p>
 
-    <ul id="history"></ul>
+    <ul id="history" ></ul>
```
Comment (ganewman42) [inline walkthroughs/week-3-server/examples/subtraction-game/src/main/webapp/index.html]: This might be an accidental change.

File: walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index-w3s2.html [hunk 14]
```diff
@@ -3,14 +3,14 @@
   <head>
     <meta charset="UTF-8">
     <title>Random Quotes</title>
-    <link rel="stylesheet" href="style.css">
-    <script src="script.js"></script>
+    <link rel="stylesheet" href="style-e.css">
+    <script src="script-e.js"></script>
```
Comment (ganewman42) [inline walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index-w3s2.html]: This looks like an error, you have script.js not script-e.js.

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    const params = new URLSearchParams();
+    params.append('maxComments', maxComments);
+    
+    let response = await fetch('/list-comments', {method: 'POST', body: params});
+    console.log("Got respone");
+    let json = await response.json();
+    console.log(json);
+    let container = document.getElementById("comment-container");
+    container.innerHTML = "";
+    for (let i = 0; i < json.length; ++i) {
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js:35]: This is well thought out - it should achieve a good layout without excessive code complexity.

**Review by ganewman42 — COMMENTED** [review]
Looks good, you've added some really nice features. I've commented with some minor cleanups, and will follow up about the async calls, which might not be necessary for your use case.

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (hieudan225) [reply portfolio/src/main/webapp/script.js]: After a bit of research, I thought that the front end validation was necessary because otherwise invalid data would be sent to the server and exceptions would be throw out all the times. On the other hand, I also thought back end validation was necessary because otherwise, evil people would be able to get around the front end validation and send hazardous data to the server. On this post (https://www.quora.com/Should-I-sanitize-inputs-clientside-or-serverside), I've seen similar arguments as well. So that was my thought, but it is easy to mess up if both sides aren't in sync like you said. What is your thought?

File: portfolio/src/main/java/com/google/sps/servlets/ListComments.java (hunk 8 above)
Comment (ganewman42) [reply portfolio/src/main/java/com/google/sps/servlets/ListComments.java]: I think you missed this one, please take a look.

File: walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index-w3s2.html (hunk 14 above)
Comment (ganewman42) [reply walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index-w3s2.html]: Could you clarify whether this is a mistake?

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (ganewman42) [reply portfolio/src/main/webapp/script.js:9]: Okay I did some more reading and talked it over with Chris as well. What I read backs up that there's probably no benefit to having these functions be async, so I'd advocate to make them synchronous again unless you want to point out some additional value that I'm missing.

**Review by ganewman42 — COMMENTED** [review]
Few remaining cleanups from my first round of comments, but thank you for addressing most of the items.

File: walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index-w3s2.html (hunk 14 above)
Comment (hieudan225) [reply walkthroughs/week-3-server/examples/random-quotes/src/main/webapp/index-w3s2.html]: Oh I missed this comment. Yes, it is a mistake.

File: portfolio/src/main/java/com/google/sps/servlets/ListComments.java (hunk 8 above)
Comment (hieudan225) [reply portfolio/src/main/java/com/google/sps/servlets/ListComments.java]: This is weird, I think I have changed this method on the previous push. 
image

File: portfolio/src/main/java/com/google/sps/data/LoginStatus.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package com.google.sps.data;
+
+public class LoginStatus {
+
+    // if login is true, the url is for logoff, else url is for login
+    public boolean login;
```
Comment (ganewman42) [inline portfolio/src/main/java/com/google/sps/data/LoginStatus.java:5]: Nit: This seems a little confusing. If the the url is for logoff when `login` is true, you might want to rename the variable to something like `isLoggedIn` or `isLogoffUrl`.

File: portfolio/src/main/java/com/google/sps/servlets/ListComments.java (hunk 8 above)
Comment (ganewman42) [reply portfolio/src/main/java/com/google/sps/servlets/ListComments.java]: Looks like the screenshot is of `DeleteAllComments`, this is `ListComments`, which I think might still have the break in it.

File: portfolio/src/main/webapp/script.js [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+
+/*Handle onload for comment section in intro.html*/
+function setMaxComments() {
+    let maxComments = document.getElementById("maxComments").value;
+
+    const params = new URLSearchParams();
+    params.append('maxComments', maxComments);
+    fetch('/max-comments', {method: 'POST', body: params});
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js]: Didn't you realize that you needed an await on the fetch since the max comments need to be set before reload?

File: portfolio/src/main/webapp/script.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    let maxCommentsRep = await fetch("/max-comments");
+    let maxCommentsText = await maxCommentsRep.text();
+    let maxComments = parseInt(maxCommentsText, 10);
+
+    
+    const params = new URLSearchParams();
+    params.append('maxComments', maxComments);
+    
+    let response = await fetch('/list-comments', {method: 'POST', body: params});
+    console.log("Got respone");
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js]: I like that you're logging for debugging! For logging statements, it's better to provide more context so that someone can understand what's happening without reading your code. For example you could log `Got response from POST to /list-comments`,  which would allow the log below to make more sense.

File: portfolio/src/main/webapp/script.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+async function deleteAllComments() {
+    let response = await fetch('/delete-all-comments', {method: 'POST'});
+    let repText = await response.text();
+    while (repText === "false") {
+        response = await fetch('/delete-all-comments', {method: 'POST'});
+        repText = await response.text();
+    }
+    console.log("Finished");
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js]: As above, something like `Finished deleting all comments` might be more helpful.

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    location.reload();
+}
+
+async function onLoad() {
+    
+    let loginStatusPromise  = await fetch("/login");
+    let loginStatus = await loginStatusPromise.json();
+    console.log(loginStatus);
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js:86]: As elsewhere, maybe `On load, login status was {whatever}`

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    if (!loginStatus.login) {
+        document.getElementById("login").style.display = "none";
+
+        const loginURL = loginStatus.url;
+        let loginURLContainer = document.getElementById("not-login");
+        let loginURLElement = document.createElement("button");
+        loginURLElement.innerText = "Click here to Login";
+        let num = 1;
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js]: This doesn't seem to be used anywhere, maybe remove it.

**Review by ganewman42 — APPROVED** [review]
A few minor cleanups, but looks good overall.

File: portfolio/src/main/java/com/google/sps/data/LoginStatus.java (hunk 16 above)
Comment (ganewman42) [reply portfolio/src/main/java/com/google/sps/data/LoginStatus.java:5]: The way you implemented it helps clarify, smart move.

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (ganewman42) [reply portfolio/src/main/webapp/script.js:9]: Out of date, resolving.

File: portfolio/src/main/webapp/script.js [hunk 22]
```diff
@@ -1,37 +1,35 @@
...
-function setMaxComments() {
+async function setMaxComments() {
     let maxComments = document.getElementById("maxComments").value;
-
     const params = new URLSearchParams();
     params.append('maxComments', maxComments);
-    fetch('/max-comments', {method: 'POST', body: params});
+    await fetch('/max-comments', {method: 'POST', body: params});
     location.reload();
+    console.log("Has set maxComments to " + maxComments);
```
Comment (ganewman42) [inline portfolio/src/main/webapp/script.js:15]: Nice clarification

**Review by ganewman42 — APPROVED** [review]
Revisions look good, thanks!
